Add some guidelines for Anchor tags

eknelson17 opened the pull request:
I regularly find myself unsure what type of link or button to use to complete certain tasks, such as showing a modal window. This PR is meant to add some guidelines when we find ourselves in those situations where basic semantics don't address the question of what is best to use.

I'm totally open to suggestions. As a team, we should decide on a few guidelines for buttons vs. inputs vs. links and when to use each one.

eknelson17 committed:
anchors: Add some guidelines on what style to use when.

dreampopclub commented on README.md:
```diff
@@ -105,4 +105,31 @@ E.g., `url`, `email`, etc.
+
+Use this style of anchor tag to create links to other pages.
+
+If using a link anchor tag as a placeholder (`<a href="#">Not a real link</a>), then be sure to add `ev.preventDefault()` to your JS click event. Otherwise, the page will jump to the top every time a user clicks your placeholder.
```
`JS click event handler.`

mhairston replied:
Double nested backticks.

mhairston replied:
Addressing multiple issues here:
- The semantic question of link vs button:
  - A button invokes an action: Save, Cancel, Exterminate.
    - Activated with the `Space` key. Screen reader users may assume this is a form submission, and be slightly more trepidacious about activiating it than a link.
    - Identified in screen readers as a button, and included in the list of form controls.
  - A link takes you somewhere (navigates).
    - Activated with the `Enter` key.
    - Identified in screen readers as a link, and included in the list of page links.
- The difference between `<button>` and `<input type="submit/button/etc">`:
  - You can use either. The advantage of `<button>` is that you can put other elements inside it, whereas you can only use plain text labels for `<input>`-style buttons.
  - A `button` is by default a _submit_ button, and will trigger the action of a form if it's inside one. Add `type="button"` to make it a non-submit button.
- A method of nullifying an href so a JS event handler can be attached without the link click reloading the page. `href="#"` by itself will just reload the page. `href="javascript://"` (there are other variations on this) will prevent the default link action.

There _are_ ambiguous cases...one common pattern is to present the primary action as a button and the secondary action as a link (e.g., Save button + Cancel link). As with everything, it's not simply a matter of style -- the semantics are important, and have real consequences for keyboard and screen reader users.

BTW, an `<a>` without the href is not new to HTML5, and they are usually just called an _anchor_. A lot of times they have an ID so you can link to a document fragment. The term "placeholder link" is new to me, and IMO less accurate than "anchor".

eknelson17 replied:
I found that term used in multiple places on a variety of sites. Seems to be used to differentiate types of anchors since a "real" link uses the same tag. If we can agree that "anchor" means `<a>` and link refers to `<a href="">`, then I'll use that. Those terms didn't seem any more common than those that I used...and I learned anchor to synonymous with link at school, not that I had the best prof for web design.

mhairston replied:
I think you learned it right: the element name itself is an abbreviation of "anchor" -- and if you have `<a href="..." id="aFragment">text</a>` that will work as both a link and an anchor. Then of course almost everyone refers to that as a "link tag," which is quite different from `<link rel="...">`

mhairston replied:
I did some extensive testing of keyboard activation of links & buttons a couple years ago. I'm trying to dig up the results, but I believe I found that across the board, links & buttons all respond to BOTH `Enter` AND `Space` on ALL browsers. The distinction above may just be convention/historical. The other thing I found is that those keys also trigger the click event, again, in ALL browsers.

I'll verify whether that's the case, and update the HTML guide.

eknelson17 replied:
We really should add some accessibility guidelines. I don't even know most of this stuff, since most of my screen reader dev experience is with desktop applications...You don't tend to have links in those.

eknelson17 replied:
Another ambiguous case that gets me regularly: A link/button that reveals hidden elements like an info modal.

Most recently I used an anchor, but I know I've used links in the past and possibly a button outside a form at one point.

dreampopclub commented:
Thanks for putting this in! Looks good.

eknelson17 committed:
Add info from @mhairston on button vs. anchor.

- Fix typos.
- Incorporate feedback and further information on the semantics being discussed.

mhairston commented on README.md:
````diff
@@ -105,4 +106,49 @@ E.g., `url`, `email`, etc.
+<a>Text</a>
+```
+
+Anchors are great for navigation or displaying hidden information on a page. Using an anchor makes the element read as clickable to a screenreader so it is the preferred solution over a `<div>` or `<span>`.
````
This is the weird thing about anchor elements: without the href, its behavior changes--it's almost a different element. Screen readers will NOT see it as clickable. And unless an event handler is attached, clicking will have no effect.

I'd consider calling the section "links", then listing anchor as a special case at the end of the link section, since it's used a lot less frequently.

mhairston commented on README.md:
```diff
@@ -105,4 +106,49 @@ E.g., `url`, `email`, etc.
+
+Anchors are great for navigation or displaying hidden information on a page. Using an anchor makes the element read as clickable to a screenreader so it is the preferred solution over a `<div>` or `<span>`.
+
+This element type first became valid in the HTML5 spec, so make sure your DOCTYPE isn't XHTML before use. If you're stuck in older HTML, use a link anchor tag.
```
This should be removed. Anchors have been usable since HTML 1.0, in every DOCTYPE.

eknelson17 committed:
Remove incorrect information

eknelson17 committed:
Update organization and examples

Reorder section to make more sense and improve readability. Add more examples and clean up existing ones.

mhairston commented on README.md:
```diff
@@ -105,4 +106,58 @@ E.g., `url`, `email`, etc.
+#### Buttons vs.Links vs. Anchors
+##### The semantic question of link vs button:
+ * A button invokes an action: Save, Cancel, Exterminate.
+  * Activated with the `Space` or `Enter` key. Screen reader users may assume this is a form submission and be slightly more trepidacious about activiating it than a link.
```
Nitpick: typo "activiating"

(it's probably _my_ typo haha)

mhairston commented on README.md:
```diff
@@ -105,4 +106,58 @@ E.g., `url`, `email`, etc.
+
+##### There _are_ ambiguous cases:
+* One common pattern is to present the primary action as a button and the secondary action as a link (e.g., Save button and Cancel link). 
+* If the button/link is meant to reveal hidden information, such as an info modal, use `<a href="#0">Show Me Stuff!</a>`.
```
Is the reason for the "#0" pointed out somewhere here?

eknelson17 replied:
Fixed 84bfa56.

eknelson17 committed:
Add #0 explanation.

...and fix a typo.
